File: macos/App/Views/Shared/Components.swift
import SwiftUI
import WayforkCore

/// Green filled / grey hollow / orange half / red cross (docs/design/02-ux.md).
struct StatusGlyphView: View {
    let glyph: StatusGlyph
    var size: CGFloat = 10

    var body: some View {
        ZStack {
            switch glyph {
            case .up:
                Circle().fill(Color.green)
            case .idle:
                Circle().stroke(Color.secondary.opacity(0.7), lineWidth: 1.2)
                    .padding(0.6)
            case .transitioning:
                Circle().stroke(Color.orange, lineWidth: 1.2).padding(0.6)
                HalfCircle().fill(Color.orange)
            case .failed:
                Circle().fill(Color.red)
                Image(systemName: "xmark")
                    .font(.system(size: size * 0.6, weight: .bold))
                    .foregroundStyle(.white)
            case .group:
                RoundedRectangle(cornerRadius: size * 0.25).fill(Color.accentColor)
            }
        }
        .frame(width: size, height: size)
        .accessibilityLabel(label)
    }

    private var label: String {
        switch glyph {
        case .up: "connected"
        case .idle: "inactive"
        case .transitioning: "connecting"
        case .failed: "failed"
        case .group: "group"
        }
    }
}

private struct HalfCircle: Shape {
    func path(in rect: CGRect) -> Path {
        var path = Path()
        let center = CGPoint(x: rect.midX, y: rect.midY)
        path.move(to: CGPoint(x: rect.midX, y: rect.minY))
        path.addArc(
            center: center, radius: rect.width / 2, startAngle: .degrees(-90),
            endAngle: .degrees(90), clockwise: false)
        path.closeSubpath()
        return path
    }
}

/// `OpenVPN` / `VLESS` pill.
struct TypeBadge: View {
    let kind: TunnelKind

    var body: some View {
        Text(StatusText.typeBadge(kind))
            .font(.system(size: 10, weight: .semibold))
            .padding(.horizontal, 5)
            .padding(.vertical, 1)
            .background(RoundedRectangle(cornerRadius: 4).fill(.quaternary))
            .foregroundStyle(.secondary)
    }
}

/// Accent marker for the tunnel handling sites without an explicit rule.
struct AccentBadge: View {
    let text: String

    var body: some View {
        Text(text)
            .font(.system(size: 10, weight: .semibold))
            .padding(.horizontal, 5)
            .padding(.vertical, 1)
            .foregroundStyle(Color.accentColor)
            .background(RoundedRectangle(cornerRadius: 4).fill(Color.accentColor.opacity(0.12)))
    }
}

/// Small rounded chip for counts and warnings (`3 rules`, `shadowed`).
struct Chip: View {
    let text: String
    var tint: Color? = nil

    var body: some View {
        Text(text)
            .font(.system(size: 11))
            .padding(.horizontal, 8)
            .padding(.vertical, 2)
            .background(Capsule().fill((tint ?? Color.secondary).opacity(0.14)))
            .foregroundStyle(tint ?? Color.secondary)
    }
}

/// White (or dark) box with a hairline border used for grouped lists.
struct GroupBackground: View {
    var body: some View {
        RoundedRectangle(cornerRadius: 6)
            .fill(Color(nsColor: .controlBackgroundColor))
            .overlay(
                RoundedRectangle(cornerRadius: 6).stroke(Color.primary.opacity(0.08), lineWidth: 1)
            )
    }
}

/// Section title of a Settings page.
struct PageTitle: View {
    let text: String

    var body: some View {
        Text(text).font(.system(size: 20, weight: .bold))
    }
}

extension View {
    /// Red focus ring used for invalid input.
    func invalidOutline(_ invalid: Bool) -> some View {
        overlay(
            RoundedRectangle(cornerRadius: 5)
                .stroke(Color.red, lineWidth: 1)
                .opacity(invalid ? 1 : 0)
        )
    }
}

/// `62 ms` in the band colour, or the red words when the tunnel is unreachable; `—` in
/// tertiary while no probe has answered yet (F14, docs/design/02-ux.md, "Variant C").
struct LatencyLabel: View {
    let sample: LatencySample?

    var body: some View {
        Group {
            if let sample, sample.unreachable {
                Text("Not reachable").fontWeight(.semibold).foregroundStyle(.red)
            } else if let sample, let milliseconds = sample.milliseconds {
                (Text("\(milliseconds)").fontWeight(.medium) + Text(" ms").font(.system(size: 10)))
                    .foregroundStyle(bandColor(LatencyBand(milliseconds: milliseconds)))
                    .monospacedDigit()
            } else {
                Text("—").foregroundStyle(.tertiary)
            }
        }
        .font(.system(size: 13))
        .lineLimit(1)
        .help(
            sample.map(LatencyFormat.tooltip)
                ?? "Measured through the tunnel every \(Int(LatencyProbe.interval)) s")
    }
}

/// Last 2 minutes of probes, 64 × 24 pt, newest at the right edge; failed probes leave a
/// dashed baseline segment (F14).
struct SparklineView: View {
    let sample: LatencySample
    nonisolated static let size = CGSize(width: 64, height: 24)
    /// Values above this are drawn at the top edge.
    nonisolated static let ceiling = 400.0

    var body: some View {
        let points = sample.history
        let latest = sample.milliseconds ?? points.compactMap { $0 }.last ?? 0
        let color = sample.unreachable ? Color.red : bandColor(LatencyBand(milliseconds: latest))
        Canvas { context, size in
            let slots = max(LatencyProbe.historyLength, 2)
            let step = size.width / CGFloat(slots - 1)
            let offset = slots - points.count
            let baseline = size.height - 2
            func point(_ index: Int, _ value: Int) -> CGPoint {
                let share = min(Double(value), Self.ceiling) / Self.ceiling
                return CGPoint(
                    x: CGFloat(index + offset) * step,
                    y: baseline - CGFloat(share) * (size.height - 4))
            }
            var line = Path()
            var gaps = Path()
            var open = false
            for (index, value) in points.enumerated() {
                if let value {
                    let p = point(index, value)
                    if open { line.addLine(to: p) } else { line.move(to: p) }
                    open = true
                } else {
                    open = false
                    let x = CGFloat(index + offset) * step
                    gaps.move(to: CGPoint(x: x, y: baseline))
                    gaps.addLine(to: CGPoint(x: min(x + step, size.width), y: baseline))
                }
            }
            context.stroke(
                line, with: .color(color),
                style: StrokeStyle(lineWidth: 1.5, lineCap: .round, lineJoin: .round))
            context.stroke(
                gaps, with: .color(.secondary.opacity(0.5)),
                style: StrokeStyle(lineWidth: 1, dash: [2, 2]))
            if let last = points.last, let value = last {
                let p = point(points.count - 1, value)
                context.fill(
                    Path(ellipseIn: CGRect(x: p.x - 1.6, y: p.y - 1.6, width: 3.2, height: 3.2)),
                    with: .color(color))
            }
        }
        .frame(width: Self.size.width, height: Self.size.height)
    }
}

/// One member line under a group card or in the expanded group (F16): dot, name, the
/// `✓ in use` / `skipped — …` note, latency right-aligned.
struct GroupMemberRowView: View {
    let row: GroupMemberRow
    var showsLatency = true

    var body: some View {
        HStack(spacing: 6) {
            StatusGlyphView(glyph: row.glyph, size: 8)
            Text(row.tunnel.name).lineLimit(1)
            if !row.note.isEmpty {
                Text(row.note)
                    .foregroundStyle(row.isActive ? Color.accentColor : Color.secondary)
                    .lineLimit(1)
            }
            Spacer(minLength: 4)
            if showsLatency {
                LatencyLabel(sample: row.latency).font(.system(size: 11))
            }
        }
        .font(.system(size: 11))
    }
}

/// Text colours for the latency bands: darker than the status dots so they read on white,
/// lighter in dark mode (the prototype's `--okt` / `--warnt` / `--badt`).
func bandColor(_ band: LatencyBand) -> Color {
    Color(
        nsColor: NSColor(name: nil) { appearance in
            let dark = appearance.bestMatch(from: [.darkAqua, .aqua]) == .darkAqua
            switch band {
            case .good:
                return dark
                    ? NSColor(red: 0.19, green: 0.82, blue: 0.35, alpha: 1)
                    : NSColor(red: 0.12, green: 0.56, blue: 0.24, alpha: 1)
            case .fair:
                return dark
                    ? NSColor(red: 1.0, green: 0.70, blue: 0.25, alpha: 1)
                    : NSColor(red: 0.66, green: 0.35, blue: 0.0, alpha: 1)
            case .poor:
                return dark
                    ? NSColor(red: 1.0, green: 0.41, blue: 0.38, alpha: 1)
                    : NSColor(red: 0.82, green: 0.20, blue: 0.17, alpha: 1)
            }
        })
}

File: windows/app/lib/app/ui/widgets/components.dart
import 'dart:math';

import 'package:fluent_ui/fluent_ui.dart';
import 'package:wayfork/core/app/latency_format.dart';
import 'package:wayfork/core/app/status_text.dart';
import 'package:wayfork/core/app/traffic_format.dart';
import 'package:wayfork/core/ipc/payloads.dart';
import 'package:wayfork/core/model/tunnel.dart';

/// Green filled / grey hollow / orange half / red cross
/// (docs/design/02-ux.md, "Status glyphs"), painted rather than composed from
/// icons so the four states line up on the same 10 px baseline.
class StatusGlyphView extends StatelessWidget {
  const StatusGlyphView({required this.glyph, this.size = 10, super.key});

  final StatusGlyph glyph;
  final double size;

  @override
  Widget build(BuildContext context) {
    final resources = FluentTheme.of(context).resources;
    return Semantics(
      label: switch (glyph) {
        StatusGlyph.up => 'connected',
        StatusGlyph.idle => 'inactive',
        StatusGlyph.transitioning => 'connecting',
        StatusGlyph.failed => 'failed',
        StatusGlyph.group => 'group',
      },
      child: CustomPaint(
        size: Size.square(size),
        painter: _GlyphPainter(
          glyph: glyph,
          up: resources.systemFillColorSuccess,
          idle: resources.textFillColorTertiary,
          transitioning: resources.systemFillColorCaution,
          failed: resources.systemFillColorCritical,
          cross: resources.textOnAccentFillColorPrimary,
          accent: FluentTheme.of(context).accentColor,
        ),
      ),
    );
  }
}

class _GlyphPainter extends CustomPainter {
  const _GlyphPainter({
    required this.glyph,
    required this.up,
    required this.idle,
    required this.transitioning,
    required this.failed,
    required this.cross,
    required this.accent,
  });

  final StatusGlyph glyph;
  final Color up;
  final Color idle;
  final Color transitioning;
  final Color failed;
  final Color cross;
  final Color accent;

  @override
  void paint(Canvas canvas, Size size) {
    final center = Offset(size.width / 2, size.height / 2);
    final radius = size.width / 2;
    switch (glyph) {
      case StatusGlyph.up:
        canvas.drawCircle(center, radius, Paint()..color = up);
      case StatusGlyph.idle:
        canvas.drawCircle(center, radius - 0.6, _stroke(idle));
      case StatusGlyph.transitioning:
        canvas.drawCircle(center, radius - 0.6, _stroke(transitioning));
        // The right half filled, as on macOS: readable at 10 px, unlike a
        // spinner.
        canvas.drawArc(
          Rect.fromCircle(center: center, radius: radius - 0.6),
          -pi / 2,
          pi,
          true,
          Paint()..color = transitioning,
        );
      case StatusGlyph.failed:
        canvas.drawCircle(center, radius, Paint()..color = failed);
        final arm = radius * 0.42;
        final pen = Paint()
          ..color = cross
          ..strokeWidth = 1.3
          ..strokeCap = StrokeCap.round;
        canvas.drawLine(
          center - Offset(arm, arm),
          center + Offset(arm, arm),
          pen,
        );
        canvas.drawLine(
          center - Offset(arm, -arm),
          center + Offset(arm, -arm),
          pen,
        );
      case StatusGlyph.group:
        // Accent square instead of the status dot (F16).
        canvas.drawRRect(
          RRect.fromRectAndRadius(
            Rect.fromCenter(
              center: center,
              width: size.width,
              height: size.height,
            ),
            Radius.circular(size.width * 0.25),
          ),
          Paint()..color = accent,
        );
    }
  }

  Paint _stroke(Color color) => Paint()
    ..color = color
    ..style = PaintingStyle.stroke
    ..strokeWidth = 1.2;

  @override
  bool shouldRepaint(_GlyphPainter old) =>
      old.glyph != glyph ||
      old.up != up ||
      old.failed != failed ||
      old.accent != accent;
}

/// `OpenVPN` / `VLESS` pill.
class TypeBadge extends StatelessWidget {
  const TypeBadge({required this.kind, super.key});

  final TunnelKind kind;

  @override
  Widget build(BuildContext context) {
    final theme = FluentTheme.of(context);
    return Container(
      padding: const EdgeInsets.symmetric(horizontal: 6, vertical: 1),
      decoration: BoxDecoration(
        color: theme.resources.controlAltFillColorSecondary,
        borderRadius: BorderRadius.circular(4),
      ),
      child: Text(
        StatusText.typeBadge(kind),
        style: theme.typography.caption?.copyWith(
          color: theme.resources.textFillColorSecondary,
          fontWeight: FontWeight.w600,
        ),
      ),
    );
  }
}

/// `↓ 1.2 MB/s ↑ 85 KB/s` with the session totals as the tooltip; `↓ — ↑ —`
/// without a fresh sample (F9). Tabular figures keep the row from jittering.
class RateLabel extends StatelessWidget {
  const RateLabel({required this.counters, super.key});

  final TrafficCounters? counters;

  @override
  Widget build(BuildContext context) {
    final theme = FluentTheme.of(context);
    final idle = counters == null || counters!.isIdle;
    return Tooltip(
      message: counters == null
          ? TrafficFormat.staleTooltip
          : TrafficFormat.tooltip(counters!),
      child: Text(
        TrafficFormat.rateLabel(counters),
        maxLines: 1,
        style: theme.typography.caption?.copyWith(
          color: idle
              ? theme.resources.textFillColorTertiary
              : theme.resources.textFillColorSecondary,
          fontFeatures: const [FontFeature.tabularFigures()],
        ),
      ),
    );
  }
}

/// The white (or dark) box with a hairline border the grouped lists sit in.
class GroupCard extends StatelessWidget {
  const GroupCard({required this.children, super.key});

  /// Rows; a divider is drawn between them.
  final List<Widget> children;

  @override
  Widget build(BuildContext context) {
    final resources = FluentTheme.of(context).resources;
    return Container(
      decoration: BoxDecoration(
        color: resources.cardBackgroundFillColorDefault,
        border: Border.all(color: resources.cardStrokeColorDefault),
        borderRadius: BorderRadius.circular(6),
      ),
      clipBehavior: Clip.antiAlias,
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          for (var index = 0; index < children.length; index++) ...[
            if (index > 0)
              Divider(
                style: DividerThemeData(horizontalMargin: EdgeInsets.zero),
              ),
            children[index],
          ],
        ],
      ),
    );
  }
}

/// The `Dashboard` / `Tunnels` heading of a page.
class PageTitle extends StatelessWidget {
  const PageTitle(this.text, {super.key});

  final String text;

  @override
  Widget build(BuildContext context) =>
      Text(text, style: FluentTheme.of(context).typography.subtitle);
}

/// Secondary body text — the shade every hint, detail line and caption uses.
class SecondaryText extends StatelessWidget {
  const SecondaryText(
    this.text, {
    this.color,
    this.maxLines = 1,
    this.overflow = TextOverflow.ellipsis,
    super.key,
  });

  final String text;
  final Color? color;
  final int? maxLines;
  final TextOverflow overflow;

  @override
  Widget build(BuildContext context) {
    final theme = FluentTheme.of(context);
    return Text(
      text,
      maxLines: maxLines,
      overflow: overflow,
      style: theme.typography.caption?.copyWith(
        color: color ?? theme.resources.textFillColorSecondary,
      ),
    );
  }
}

/// A yes/no dialog: the Win32 counterpart of the macOS `Alerts.show` with two
/// buttons. True when the user picked [confirm].
Future<bool> showQuestionDialog(
  BuildContext context, {
  required String title,
  required String message,
  required String confirm,
  String cancel = 'Cancel',
  bool destructive = false,
}) async {
  final answer = await showDialog<bool>(
    context: context,
    builder: (context) => ContentDialog(
      constraints: const BoxConstraints(maxWidth: 460),
      title: Text(title),
      content: Text(message),
      actions: [
        Button(
          onPressed: () => Navigator.of(context).pop(false),
          child: Text(cancel),
        ),
        FilledButton(
          style: destructive
              ? ButtonStyle(
                  backgroundColor: WidgetStatePropertyAll(
                    FluentTheme.of(context).resources.systemFillColorCritical,
                  ),
                )
              : null,
          onPressed: () => Navigator.of(context).pop(true),
          child: Text(confirm),
        ),
      ],
    ),
  );
  return answer ?? false;
}

/// A small rounded label next to a name: `2 rules`, `paused`, `shadowed`.
/// [tint] colours the text and the background of the ones that carry a
/// warning; the tooltip says why.
class Chip extends StatelessWidget {
  const Chip(this.text, {this.tint, this.tooltip, super.key});

  final String text;
  final Color? tint;
  final String? tooltip;

  @override
  Widget build(BuildContext context) {
    final theme = FluentTheme.of(context);
    final tint = this.tint;
    final chip = Container(
      padding: const EdgeInsets.symmetric(horizontal: 6, vertical: 1),
      decoration: BoxDecoration(
        color: tint == null
            ? theme.resources.controlAltFillColorSecondary
            : tint.withValues(alpha: 0.16),
        borderRadius: BorderRadius.circular(4),
      ),
      child: Text(
        text,
        maxLines: 1,
        style: theme.typography.caption?.copyWith(
          color: tint ?? theme.resources.textFillColorSecondary,
        ),
      ),
    );
    final message = tooltip;
    return message == null ? chip : Tooltip(message: message, child: chip);
  }
}

/// A field with the validation message the model returned under it.
class FieldWithError extends StatelessWidget {
  const FieldWithError({required this.child, this.error, super.key});

  final Widget child;
  final String? error;

  @override
  Widget build(BuildContext context) => Column(
    crossAxisAlignment: CrossAxisAlignment.start,
    children: [
      child,
      if (error case final message?)
        Padding(
          padding: const EdgeInsets.only(top: 2),
          child: SecondaryText(
            message,
            color: FluentTheme.of(context).resources.systemFillColorCritical,
            maxLines: 2,
          ),
        ),
    ],
  );
}

/// Addresses, hashes and URIs: fixed width so they line up and do not reflow.
class MonoText extends StatelessWidget {
  const MonoText(this.text, {super.key});

  final String text;

  @override
  Widget build(BuildContext context) {
    final theme = FluentTheme.of(context);
    return Text(
      text,
      maxLines: 1,
      overflow: TextOverflow.ellipsis,
      style: theme.typography.caption?.copyWith(
        fontFamily: 'Consolas',
        color: theme.resources.textFillColorSecondary,
      ),
    );
  }
}

/// Accent marker next to a name: `Default`, `Group`.
class AccentBadge extends StatelessWidget {
  const AccentBadge(this.text, {super.key});

  final String text;

  @override
  Widget build(BuildContext context) {
    final theme = FluentTheme.of(context);
    return Container(
      padding: const EdgeInsets.symmetric(horizontal: 6, vertical: 1),
      decoration: BoxDecoration(
        color: theme.accentColor.withValues(alpha: 0.12),
        borderRadius: BorderRadius.circular(4),
      ),
      child: Text(
        text,
        style: theme.typography.caption?.copyWith(
          color: theme.accentColor,
          fontWeight: FontWeight.w600,
        ),
      ),
    );
  }
}

/// Text colours for the latency bands (F14): darker than the status dots so
/// they read on white, lighter in dark mode.
Color bandColor(BuildContext context, LatencyBand band) {
  final dark = FluentTheme.of(context).brightness == Brightness.dark;
  return switch (band) {
    LatencyBand.good =>
      dark ? const Color(0xFF30D158) : const Color(0xFF1F8F3D),
    LatencyBand.fair =>
      dark ? const Color(0xFFFFB340) : const Color(0xFFA85900),
    LatencyBand.poor =>
      dark ? const Color(0xFFFF6961) : const Color(0xFFD1332B),
  };
}

/// `62 ms` in the band colour, or the red words when the tunnel is unreachable;
/// `—` while no probe has answered yet (F14).
class LatencyLabel extends StatelessWidget {
  const LatencyLabel({required this.sample, this.fontSize = 13, super.key});

  final LatencySample? sample;
  final double fontSize;

  @override
  Widget build(BuildContext context) {
    final theme = FluentTheme.of(context);
    final sample = this.sample;
    final Widget child;
    if (sample != null && sample.unreachable) {
      child = Text(
        'Not reachable',
        style: TextStyle(
          fontSize: fontSize,
          fontWeight: FontWeight.w600,
          color: theme.resources.systemFillColorCritical,
        ),
      );
    } else if (sample != null && sample.milliseconds != null) {
      final color = bandColor(context, LatencyBand.of(sample.milliseconds!));
      child = Text.rich(
        TextSpan(
          children: [
            TextSpan(
              text: '${sample.milliseconds}',
              style: TextStyle(fontWeight: FontWeight.w500, color: color),
            ),
            TextSpan(
              text: ' ms',
              style: TextStyle(fontSize: fontSize - 3, color: color),
            ),
          ],
        ),
        style: TextStyle(
          fontSize: fontSize,
          fontFeatures: const [FontFeature.tabularFigures()],
        ),
      );
    } else {
      child = Text(
        '—',
        style: TextStyle(
          fontSize: fontSize,
          color: theme.resources.textFillColorTertiary,
        ),
      );
    }
    return Tooltip(
      message: sample == null
          ? 'Measured through the tunnel every ${LatencyProbe.interval.inSeconds} s'
          : LatencyFormat.tooltip(sample),
      child: child,
    );
  }
}

/// Last 2 minutes of probes, 64 × 24 px, newest at the right edge; failed
/// probes leave a dashed baseline segment (F14).
class SparklineView extends StatelessWidget {
  const SparklineView({required this.sample, super.key});

  final LatencySample sample;
  static const size = Size(64, 24);
  static const ceiling = 400.0;

  @override
  Widget build(BuildContext context) {
    final latest =
        sample.milliseconds ?? sample.history.whereType<int>().lastOrNull ?? 0;
    final color = sample.unreachable
        ? FluentTheme.of(context).resources.systemFillColorCritical
        : bandColor(context, LatencyBand.of(latest));
    return CustomPaint(
      size: size,
      painter: _SparklinePainter(
        points: sample.history,
        color: color,
        gap: FluentTheme.of(context).resources.textFillColorTertiary,
      ),
    );
  }
}

class _SparklinePainter extends CustomPainter {
  const _SparklinePainter({
    required this.points,
    required this.color,
    required this.gap,
  });

  final List<int?> points;
  final Color color;
  final Color gap;

  @override
  void paint(Canvas canvas, Size size) {
    final slots = max(LatencyProbe.historyLength, 2);
    final step = size.width / (slots - 1);
    final offset = slots - points.length;
    final baseline = size.height - 2;
    Offset point(int index, int value) {
      final share =
          min(value.toDouble(), SparklineView.ceiling) / SparklineView.ceiling;
      return Offset(
        (index + offset) * step,
        baseline - share * (size.height - 4),
      );
    }

    final line = Path();
    var open = false;
    final gapPaint = Paint()
      ..color = gap.withValues(alpha: 0.5)
      ..strokeWidth = 1;
    for (var index = 0; index < points.length; index++) {
      final value = points[index];
      if (value != null) {
        final p = point(index, value);
        if (open) {
          line.lineTo(p.dx, p.dy);
        } else {
          line.moveTo(p.dx, p.dy);
        }
        open = true;
      } else {
        open = false;
        final x = (index + offset) * step;
        final end = min(x + step, size.width);
        // A dashed segment: two-pixel dashes along the baseline.
        for (var dash = x; dash < end; dash += 4) {
          canvas.drawLine(
            Offset(dash, baseline),
            Offset(min(dash + 2, end), baseline),
            gapPaint,
          );
        }
      }
    }
    canvas.drawPath(
      line,
      Paint()
        ..color = color
        ..style = PaintingStyle.stroke
        ..strokeWidth = 1.5
        ..strokeCap = StrokeCap.round
        ..strokeJoin = StrokeJoin.round,
    );
    final last = points.lastOrNull;
    if (last != null) {
      canvas.drawCircle(
        point(points.length - 1, last),
        1.6,
        Paint()..color = color,
      );
    }
  }

  @override
  bool shouldRepaint(_SparklinePainter old) =>
      old.points != points || old.color != color;
}

/// One member line under a group card or in the expanded group (F16): dot,
/// name, the `✓ in use` / `skipped — …` note, latency right-aligned.
class GroupMemberRowView extends StatelessWidget {
  const GroupMemberRowView({
    required this.row,
    this.showsLatency = true,
    super.key,
  });

  final GroupMemberRow row;
  final bool showsLatency;

  @override
  Widget build(BuildContext context) {
    final theme = FluentTheme.of(context);
    return Row(
      children: [
        StatusGlyphView(glyph: row.glyph, size: 8),
        const SizedBox(width: 6),
        Text(row.tunnel.name, style: theme.typography.caption),
        if (row.note.isNotEmpty) ...[
          const SizedBox(width: 6),
          Flexible(
            child: Text(
              row.note,
              maxLines: 1,
              overflow: TextOverflow.ellipsis,
              style: theme.typography.caption?.copyWith(
                color: row.isActive
                    ? theme.accentColor
                    : theme.resources.textFillColorSecondary,
              ),
            ),
          ),
        ],
        const Spacer(),
        if (showsLatency) LatencyLabel(sample: row.latency, fontSize: 11),
      ],
    );
  }
}
